pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionHandler.java

```diff
@@ -66,20 +67,21 @@ protected void grabCnx() {
             return;
         }
 
-        synchronized (this) {
-            if (duringConnect) {
-                log.info("[{}] [{}] Skip grabbing the connection since there is a pending connection",
-                        state.topic, state.getHandlerName());
-                return;
-            }
-            if (isValidStateForReconnection()) {
-                duringConnect = true;
-            } else {
-                // Ignore connection closed when we are shutting down
-                log.info("[{}] [{}] Ignoring reconnection request (state: {})",
-                        state.topic, state.getHandlerName(), state.getState());
+        if (duringConnect.get()) {
+            log.info("[{}] [{}] Skip grabbing the connection since there is a pending connection",
+                    state.topic, state.getHandlerName());
+            return;
+        }
+        if (isValidStateForReconnection()) {
+            if (!duringConnect.compareAndSet(false, true)) {
```

Comment:
nit: avoid `!` and swap the block - my experience is that double negative predicts are brittle.

N.B. "nit" means it's not a blocker to this patch but nice to have IMO.

Reply:
Good suggestion. I use two early return code blocks now.